File: app/json_cleaner.js
const fs = require("fs");

// Try to parse string values as JSON
function tryParse(value) {
  if (typeof value === "string") {
    try {
      return JSON.parse(value); // removes \n, \", etc
    } catch (e) {
      return value;
    }
  }
  return value;
}

// Clean JSON structure
function cleanJsonString(data) {
  if (Array.isArray(data)) {
    return data.map(item => {
      const cleanedItem = {};
      for (const key in item) {
        cleanedItem[key] = tryParse(item[key]);
      }
      return cleanedItem;
    });
  } else if (typeof data === "object" && data !== null) {
    const cleanedObj = {};
    for (const key in data) {
      cleanedObj[key] = tryParse(data[key]);
    }
    return cleanedObj;
  } else {
    return data;
  }
}

// Read JSON file
function readJsonFile(filePath) {
  const raw = fs.readFileSync(filePath, "utf-8");
  return JSON.parse(raw);
}

// Main execution
const inputFile = "books/n8n-chatgpt-output.json";
const outputFile = "books/n8n-chatgpt-output-cleaned.json";

const data = readJsonFile(inputFile);
const cleanedData = cleanJsonString(data);

fs.writeFileSync(outputFile, JSON.stringify(cleanedData, null, 2), "utf-8");

console.log(`Cleaned data written to ${outputFile}`);

File: app/json_cleaner.py
import json

def clean_json_string(data):
    """
    Takes a dict/list and parses any string fields that contain JSON
    """
    def try_parse(value):
        if isinstance(value, str):
            try:
                return json.loads(value)  # removes \n, \", etc properly
            except:
                return value
        return value

    if isinstance(data, list):
        return [{k: try_parse(v) for k, v in item.items()} for item in data]
    elif isinstance(data, dict):
        return {k: try_parse(v) for k, v in data.items()}
    else:
        return data
    
def read_json_file(file_path):
    with open(file_path, "r") as f:
        data = json.load(f)  # parses JSON into Python object
    return data

data = read_json_file("books/n8n-chatgpt-output.json")
cleaned_data = clean_json_string(data)

output_file = "books/n8n-chatgpt-output-cleaned.json"
with open(output_file, "w", encoding="utf-8") as f:
    json.dump(cleaned_data, f, indent=2, ensure_ascii=False)

print(f"Cleaned data written to {output_file}")